### PR TITLE
Add automatic shape inference.

### PR DESCRIPTION
**Pull request by matsuyamax**
### Simple NN:
```python
model = Sequential()
model.add(Dense(64, input_dim=20))
model.add(Dropout(0.5))
model.add(Dense(64))

# alternatively
model = Sequential()
model.add(Dense(64, input_shape=(20,)))
model.add(Dropout(0.5))
model.add(Dense(64))

```

### Recurrent network:
```python
model = Sequential()
model.add(Embedding(max_features, 256))
model.add(LSTM(128))

# with LSTM as first layer
model = Sequential()
model.add(LSTM(128, input_dim=32))

# with LSTM as first layer, alternative
model = Sequential()
model.add(LSTM(128, input_shape=(None, 32)))

# adding sequence length information
model = Sequential()
model.add(LSTM(128, input_dim=32, input_length=10, return_sequences=True))
model.add(Flatten())
model.add(Dense(16)) # now this works automatically

# or alternatively
model = Sequential()
model.add(LSTM(128, input_shape=(10, 32), return_sequences=True))

# also possible with Embedding, TimeDistributedDense, Convolution1D
model = Sequential()
model.add(Embedding(max_features, 256, input_length=10))
model.add(LSTM(128, return_sequences=True))
```

### Convnet:
```python
model = Sequential()
# input: 100x100 images with 3 channels -> (3, 100, 100) tensors.
# this applies 32 convolution filters of size 3x3 each.
model.add(Convolution2D(32, 3, 3, border_mode='full', input_shape=(3, 100, 100))) 
model.add(Activation('relu'))
model.add(Convolution2D(32, 3, 3))
model.add(Activation('relu'))
model.add(MaxPooling2D(poolsize=(2, 2)))
model.add(Dropout(0.25))

model.add(Convolution2D(64, 3, 3, border_mode='valid')) 
model.add(Activation('relu'))
model.add(Convolution2D(64, 3, 3)) 
model.add(Activation('relu'))
model.add(MaxPooling2D(poolsize=(2, 2)))
model.add(Dropout(0.25))

model.add(Flatten())
model.add(Dense(256))  # boom
model.add(Activation('relu'))
model.add(Dropout(0.5))

model.add(Dense(10))
model.add(Activation('softmax'))
```

### Graph:
```python
graph = Graph()
graph.add_input(name='input1', input_shape=(32,))
graph.add_input(name='input2', input_shape=(32,))
graph.add_node(Dense(16), name='dense1', input='input1')
graph.add_node(Dense(4), name='dense2', input='input2')
graph.add_node(Dense(4), name='dense3', input='dense1')
graph.add_output(name='output', inputs=['dense2', 'dense3'], merge_mode='sum')
graph.compile('rmsprop', {'output':'mse'})
```

- All examples are updated.
- All documentation is updated.
- All tests are updated.

## Moreover:
- faster convolutional layers (by providing `image_shape` and `filter_shape` to Theano).
- multiple bugs in the library and its tests fixed, additional correctness checks added
- small API consistency fixes.

### CODE REVIEW

File: keras/layers/containers.py [hunk 1]
```diff
@@ -167,12 +166,14 @@ def get_output(self, train=False):
...
             return dict([(k, v.get_output(train)) for k, v in self.outputs.items()])
 
-    def add_input(self, name, ndim=2, dtype='float'):
+    def add_input(self, name, input_shape, dtype='float'):
         if name in self.namespace:
             raise Exception('Duplicate node identifier: ' + name)
         self.namespace.add(name)
         self.input_order.append(name)
         layer = Layer()  # empty layer
+        layer.set_input_shape(input_shape)
```
Comment (fchollet) [inline keras/layers/containers.py]: Bit of redundancy: this sets `input` twice (here and in the next few lines).

File: keras/layers/core.py [hunk 2]
```diff
@@ -323,18 +362,20 @@ def set_weights(self, weights):
...
-        return {"name": self.__class__.__name__,
-                "layers": [l.get_config() for l in self.layers],
-                "mode": self.mode,
-                "concat_axis": self.concat_axis}
+        config = {"name": self.__class__.__name__,
+                  "layers": [l.get_config() for l in self.layers],
+                  "mode": self.mode,
+                  "concat_axis": self.concat_axis}
+        base_config = super(Merge, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
```
Comment (fchollet) [inline keras/layers/core.py]: We might need a more standard way to merge dictionaries (with the latter dict taking precedence). Like a `config` decorator. Or maybe that would introduce too much implicitness.

File: keras/layers/core.py [hunk 3]
```diff
@@ -471,59 +516,65 @@ def get_output(self, train=False):
...
-
-        if name is not None:
-            self.set_name(name)
-
-    def set_name(self, name):
-        self.W.name = '%s_W' % name
-        self.b.name = '%s_b' % name
+        if self.initial_weights is not None:
+            self.set_weights(self.initial_weights)
+            del self.initial_weights
```
Comment (fchollet) [inline keras/layers/core.py]: The weight setting boilerplate code is repeated in all layers, would there be a way to abstract it?

**Comment by fchollet** [comment]
A more elegant API, for a more civilized age. 

Thanks for the excellent work. This is a major UX win.

Now let's break everyone's code. Merging : )

**Comment by rushter** [comment]
There is a typo in your example.
`model.add(MaxPooling2D(poolsize=(2, 2)))`
Should be:
`model.add(MaxPooling2D(pool_size=(2, 2)))`

**Comment by ipod825** [comment]
It is possible to add an add_input function to the Sequential model. In such way, we could remove all the input_shape or input_dim argument from all layers' constructor.

**Pull request merged**